airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py

```diff
@@ -1106,20 +1137,123 @@ def add_to_outputs(
+                        {{ '#}' }}
+                        {{ '{%' }}
+                        if final_table_relation is not none and {{ quoted_unique_key }} in adapter.get_columns_in_relation(final_table_relation)|map(attribute='name')
+                        {{ '%}' }}
+
+                        -- Delete records which are no longer active:
+                        -- 1. Find the records which are being updated by querying the _scd_new_data model
+                        -- 2. Then join that against the SCD model to find the records which have no row with _airbyte_active_row = 1
+                        -- We can't just delete all the modified_ids from final_table because those records might still be active, but not included
+                        -- in the most recent increment (i.e. the final table model would not re-insert them, so the data would be incorrectly lost).
+                        -- In fact, there's no guarantee that the active record is included in the previous_active_scd_data CTE either,
+                        -- so we _must_ join against the entire SCD table to find the active row for each record.
+                        -- We're using a subquery because not all destinations support CTEs in DELETE statements (c.f. Snowflake).
+                        -- Similarly, the subquery doesn't use CTEs because Clickhouse doesn't support CTEs inside delete conditions.
+                        {{ delete_statement }} in (
+                            select modified_ids.{{ unique_key }}
+                            from
+                                (
+                                    select nullif(scd_table.{{ unique_key }}, '') as {{ unique_key }} from {{ '{{ this }}' }} scd_table
+-- TODO is this even necessary?
+--                              inner join modified_ids on scd_table.{{ unique_key }} = modified_ids.{{ unique_key }}
+                                    where {{ active_row_column_name }} =  1
+                                ) scd_active_rows
+                                right outer join (
+                                    select
+                                        {{ '{{' }} dbt_utils.surrogate_key([
+                                            {%- for primary_key in primary_keys %}
+                                                {{ primary_key }},
+                                            {%- endfor %}
+                                        ]) {{ '}}' }} as {{ unique_key }}
+                                    from {{ quoted_scd_new_data_table }}
+                                    where 1=1
+                                        {{ incremental_clause }}
+                                ) modified_ids
+                                on modified_ids.{{ unique_key }} = scd_active_rows.{{ unique_key }}
+                            group by modified_ids.{{ unique_key }}
+                            having count(scd_active_rows.{{ unique_key }}) = 0
```

Comment:
Can this be simplified??

I would feel more comfortable if we could do a different `incremental_clause` than a complex join to identify what needs to be dropped?


For example, can we maybe use the fact that we have a `normalized_at` column and identify all the freshly written rows from scd tables (they have been written after `max(normalized_at)` column of the final table)

We don't have a reason to stick to `airbyte_emitted_at` columns only?

Reply:
lol. I was actually trying to avoid that - I feel like incremental logic is already complicated enough, and being incremental against two columns which function similarly but not identically seemed really confusing. I'll try spiking it out, maybe it'll turn out to be really straightforward

Reply:
that sounds much simpler!
is this still a post-hook on the scd table?

Reply:
yeah. I think ideally it would be a pre-hook on the final table, but suspect that might cause some weird behavior with repeated failures

Reply:
pushed some code just for local backup. will run a few local syncs for further testing and rerequest review when I'm confident it works